src/Rings/mpoly-graded.jl

```diff
@@ -2189,13 +2189,46 @@ function homogenization(V::Vector{T},  W::Union{ZZMatrix, Matrix{<:IntegerUnion}
   return [_homogenization(V[i], S, pos) for i=1:l]
 end
 
+
+# To homogenize an ideal we look for some additional simple elements
+# (as redundant generators), and then apply the standard algorithm.
+
+# Internal function: used only in gens_for_homog (immediately below)
+function num_terms(f)
+    return length(collect(coefficients(f)));
+end;
+
+# Internal function: used only in homogenization (immediately below)
+# This function is an "empirical hack": usually makes homogenization faster.
+# Return gens(I) and some further "small" polys in I.
+function gens_for_homog(I::MPolyIdeal{T}) where {T <: MPolyRingElem}
+    @req  !is_zero(I)  "Ideal must be non-zero"
+    @req  !is_one(I)   "Ideal must not be whole ring"
+    OrigR = parent(gens(I)[1])      # Since I is not zero, it has at least 1 gen.
+    # Next few lines: we adjoin some more small, redundant gens.
+    # !!HEURISTIC!!  We use AveNumTerms as size limit for redundant gens we shall adjoin.
+    AveNumTerms = sum([num_terms(f)  for f in gens(I)])/length(gens(I)) ## floating-point!
+    GB1 = elements(groebner_basis(I, ordering=degrevlex(OrigR)))
+    GB1 = filter(f -> (num_terms(f) < 2*AveNumTerms), GB1)
+    GB2 = elements(groebner_basis(I, ordering=deglex(OrigR)))  # or degrevlex with vars in reverse order?
+    GB2 = filter(f -> (num_terms(f) < 2*AveNumTerms), GB2)
+    return vcat(gens(I), GB1, GB2)
```

Comment:
I made some "randomly generated" tests, and it was generally faster (and only rarely a bit slower) when using these GB computations.  If you have some "real" inputs, we can test them.  In summary: the costs of these two GB computations were (usually) lower than the speed gain when computing the saturation.  Conceivably, this may change when Hans has a better implementation of saturation -- the question should then be revisited.
Of course, if you have a better way of generating further "small" elements of the ideal, let me know.

Reply:
I am considering changing the "short-cut" case to
```
 if all(is_homogeneous, gens(I)) 
        Hgens = homogenization(gens(I))
        return ideal(parent(Hgens), Hgens)
```
The zero ideal needs some special handling (which I have omitted here).

Reply:
Ah!  The weight matrix `W` is an input parameter, and not intrinsic to the graded ring... ?1?

Reply:
I do not understand. What is the short-cut case? Also, I do not understand your code. You homogenize the generators with respect to the standard ZZ-grading if they are homogeneous with respect to the standard ZZ-grading?

And which weight matrix do you mean?

Reply:
OK, maybe the "short-cut case" is useless: I meant if the given generators already happen to be homogeneous (with respect to `W`).  I'll ask Hans whether saturation by a variable which does not appear in the generators as handled efficiently... if so, then my "short-cut" is quite unnecessary.  The call to `homogenization` when the generators are already homogeneous is simply to map them into the correct ring [as you pointed out].
I am puzzled as to why the input parameters to `homogenization(I,W,var,pos)` include `W`, `var` and `pos`.  I would have expected these 3 parameters to be "instrinsic" to the graded polynomial ring.  The documentation says that `is_homogeneous` takes just 1 argument; and the examples suggest that the grading used is that of the graded polynomial ring.  So it is not clear to me why I must specify `W` when calling `homogenization`.  To some extent I can understand why `var` and `pos` are necessary (though it seems like they should be recorded inside the graded polynomial ring along with the weight matrix `W`)

Reply:
The point here is that, typically, we do not have the graded ring yet. That is, `I` is not an ideal of a graded ring . So you need to tell the function with respect to which grading it should homogenize. One method for `is_homogeneous` is Bill's AA version of `is_homogeneous` which checks whether a polynomial is homogeneous with respect to the standard ZZ-grading (so this applies also to polynomials in non-graded rings, but has a specific grading in mind). The methods `is_homogeneous`  we implemented in addition apply to elements of graded rings only. So this is another case of confusion arising from the history of AA.

Reply:
Ah!  So in practice I cannot use `is_homogeneous` on the input generators, right?  That excludes the "short-cut".
Indeed it is confusing that "homogenization" has such a context-dependent meaning.  I must review my code.
I also spoke to Hans: _Singular_ does not yet handle specially the case of saturating by variables which do not appear in the generators.